crates/client/rpc/src/errors.rs

```diff
@@ -1,56 +1,24 @@
 use jsonrpsee::types::error::{CallError, ErrorObject};
-use sp_runtime::DispatchError;
+use mp_starknet::starknet_serde::StarknetRpcApiError;
 
-#[derive(thiserror::Error, Clone, Copy, Debug)]
-pub enum StarknetRpcApiError {
-    #[error("Failed to write transaction")]
-    FailedToReceiveTxn = 1,
-    #[error("Contract not found")]
-    ContractNotFound = 20,
-    #[error("Invalid message selector")]
-    InvalidMessageSelector = 21,
-    #[error("Invalid call data")]
-    InvalidCallData = 22,
-    #[error("Block not found")]
-    BlockNotFound = 24,
-    #[error("Transaction hash not found")]
-    TxnHashNotFound = 25,
-    #[error("Invalid transaction index in a block")]
-    InvalidTxnIndex = 27,
-    #[error("Class hash not found")]
-    ClassHashNotFound = 28,
-    #[error("Class already declared")]
-    ClassAlreadyDeclared = 51,
-    #[error("Requested page size is too big")]
-    PageSizeTooBig = 31,
-    #[error("There are no blocks")]
-    NoBlocks = 32,
-    #[error("The supplied continuation token is invalid or unknown")]
-    InvalidContinuationToken = 33,
-    #[error("Contract error")]
-    ContractError = 40,
-    #[error("Invalid contract class")]
-    InvalidContractClass = 50,
-    #[error("Too many storage keys requested")]
-    ProofLimitExceeded = 10000,
-    #[error("Too many keys provided in a filter")]
-    TooManyKeysInFilter = 34,
-    #[error("Internal server error")]
-    InternalServerError = 500,
-    #[error("Failed to fetch pending transactions")]
-    FailedToFetchPendingTransactions = 38,
-    #[error("Unimplemented method")]
-    UnimplementedMethod = 501,
+pub struct TxError(pub StarknetRpcApiError);
```

Comment:
Why do we need this wrapper struct?
Can't you impl `jsonrpsee::core::Error` directly on `StarknetRpcApiError`?

Reply:
I meant implenting `jsonrpsee::core::Error` things directly where `StarknetRpcApiError` is defined

Reply:
And maybe it should not be into primitve/starknet, but rather primitve/rpc or primitive/errors

Reply:
Can't because we would need to add jsonrpsee as a dep and it breaks everything

Reply:
wdym "it break everything"?
Put it behind a feature flag may solve this

Reply:
Unfortunately not cause we also import it within the pallet `runtime_api.rs` so we can't have it as a std feature flag

Reply:
not as a std feature.
In cargo.toml:
```rust
jsonrpsee = {version = "...", optional = true}

[features]
std = [...]
jsonrpsee = ["std", "dep:jsonrpsee"]
```

Then in the file you create the `StarknetRpcApiError`:
```rust
#[cfg(feature = "jsonrpsee"]
impl From<StarknetRpcApiError> for jsonrpsee::core::Error {
    fn from(err: TxError) -> Self {
        jsonrpsee::core::Error::Call(CallError::Custom(ErrorObject::owned(err.0 as i32, err.0.to_string(), None::<()>)))
    }
}
```

This way you can get rid of the wrapper type.

Reply:
And I don't think the error type should go in `primitive/starknet`

Reply:
What I would do instead is define in pallet/runtime_api.rs a `StarknetTransactionExecutionError`:
```rust
pub enum StarknetTransactionExecutionError {
  ContractNotFound,
  ClassAlreadyDeclared,
  ClassHashNotFound,
  InvalidContractClass
}
```

I would keep my `StarknetRpcApiError` in the client, where it was `crates/client/rpc/src/errors.rs`, but add this:
```rust
StarknetRpcApiError {
   ...,
   Runtime(#[from] StarknetTransactionExecutionError)
}
```

This way you can keep your `From<StarknetRpcApiError> for jsonrpsee::core::Error` in the client, without bothering the pallet with it.